### PR TITLE
Add support for comparing related types in assertTrue in Scala 2

### LINKED ISSUE #8874: Scala 2.13 - cannot compare child class to parent class in assertTrue
Scala 2.13 only

https://scastie.scala-lang.org/rwgljZz2Q6m0aDTwkBDczA

```scala
import zio.test.assertTrue

class A
class B extends A
val a = new A()
val b = new B()
assertTrue(a == b)
assertTrue(b == a)
```

error:
```
type mismatch;
 found   : Playground.A
 required: Playground.B
not found: value location
```

Prepared test for `SmartAssertionSpec.scala`

```scala
    test("equalTo compiles when comparing different primitive types") {
      val a = 1
      val b = 1L
      assertTrue(a == b) && assertTrue(b == a)
    },
    test("equalTo compiles when comparing types that are direct descendants") {
      class A
      class B extends A
      val a = new A()
      val b = new B()
      assertTrue(a == b) && assertTrue(b == a)
    },
```

### PR DESCRIPTION
**Pull request by andrzejressel**
Fixes #8874

### CODE REVIEW

File: test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala [hunk 1]
```diff
@@ -199,11 +200,32 @@ object SmartAssertionSpec extends ZIOBaseSpec {
...
+      class B
+      val a = new A()
+      val b = new B()
+      assertTrue(a == b) && assertTrue(b == a)
+      """)
+      for {
+        result <- resultZIO
+      } yield assertTrue(result.is(_.left).contains("type mismatch"))
+    } @@ scala2Only,
     test("comparison compiles when comparing different types") {
```
Comment (hearnadam) [inline test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala]: Can you mark this one as 'primitive' as well?

**Review by hearnadam — APPROVED** [review]

File: test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala (hunk 1 above)
Comment (andrzejressel) [reply test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala]: Fixed - after this is merged I can add more tests related to comparing - comparing unrelated classes, comparing related classes etc. - I believe currently behaviour is a bit undefined there.

File: test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala [hunk 2]
```diff
@@ -199,12 +200,33 @@ object SmartAssertionSpec extends ZIOBaseSpec {
...
+      class A
+      class B
+      val a = new A()
+      val b = new B()
+      assertTrue(a == b) && assertTrue(b == a)
+      """)
+      for {
+        result <- resultZIO
+      } yield assertTrue(result.is(_.left).contains("type mismatch"))
+    } @@ scala2Only,
```
Comment (hearnadam) [inline test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala:228]: Will this not fail with 'type mismatch' on Scala 3?

Comment (andrzejressel) [reply test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala:228]: Nope:
```
  - SmartAssertionSpec / equalTo does not compile when comparing unrelated types in Scala 2 - 26 ms
    ? Either was Right
    result.is(_.left).contains("type mismatch")
    result = Right(value = ())
    at D:\MojeProgramy\zio\test-tests\shared\src\test\scala\zio\test\SmartAssertionSpec.scala:227
```
but it does not work correctly (at least the diffing part): https://github.com/zio/zio/issues/8754

**Comment by jdegoes** [comment]
@andrzejressel Would like to get this in. Please re-open when CI is passing! 🙏

**Comment by andrzejressel** [comment]
image

But it passed everything ...

**Pull request closed without merging**